### PR TITLE
Bug-Fix: parsing type strings for tuples containing static arrays of tuples

### PR DESCRIPTION
**Pull request by ahangsu**
Fix https://github.com/algorand/avm-abi/issues/12

### CODE REVIEW

File: src/main/java/com/algorand/algosdk/abi/ABIType.java [hunk 1]
```diff
@@ -118,8 +118,13 @@ else if (str.charAt(i) == ')') {
                 if (parenStack.isEmpty())
```
Comment (bbroder-algo) [inline src/main/java/com/algorand/algosdk/abi/ABIType.java:118]: similar error in the exception above in 106

File: src/test/java/com/algorand/algosdk/abi/TestTypes.java [hunk 2]
```diff
@@ -238,6 +238,22 @@ public void TestTypeFromStringValid() {
                         )
                 )
         );
+        assertThat(ABIType.valueOf("(uint32,(uint64,bool)[10],byte)")).isEqualTo(
```
Comment (bbroder-algo) [inline src/test/java/com/algorand/algosdk/abi/TestTypes.java:241]: maybe an additional one that puts the static array at the end so it tests the end-of-string condition in addition to the comma condition.

File: src/main/java/com/algorand/algosdk/abi/ABIType.java (hunk 1 above)
Comment (ahangsu) [reply src/main/java/com/algorand/algosdk/abi/ABIType.java:118]: 👍 good find

File: src/test/java/com/algorand/algosdk/abi/TestTypes.java (hunk 2 above)
Comment (ahangsu) [reply src/test/java/com/algorand/algosdk/abi/TestTypes.java:241]: got it, I will mimic the cases in avm-abi one

**Review by algochoi — APPROVED** [review]
Thanks for the fix + tests look good!

File: src/main/java/com/algorand/algosdk/abi/ABIType.java [hunk 3]
```diff
@@ -118,8 +118,13 @@ else if (str.charAt(i) == ')') {
...
                     throw new IllegalArgumentException("parsing error: tuple parentheses are not balanced: " + str);
                 int leftParenIndex = parenStack.pop();
-                if (parenStack.isEmpty())
+                if (parenStack.isEmpty()) {
+                    int forwardIndex = i + 1;
```
[the remark this answers is not in the document]
Comment (ahangsu) [reply src/main/java/com/algorand/algosdk/abi/ABIType.java:122]: 👍 will do

**Review by michaeldiamant — APPROVED** [review]
@ahangsu Thanks for supplying a fix - As intended, I confirmed:
* Added tests fail prior to the PR.
* Implementation mirrors other SDKs.

Out-of-scope passing remark:  I can imagine building a property-based test as a way to build more confidence in test case generation.  Though it's a big enough effort that I don't think it's worth pursuing in-scope to the story.

**Pull request merged**